sklearn/mixture/_gaussian_mixture_ic.py

```diff
@@ -0,0 +1,796 @@
+                affinity.remove("cosine")
+                warnings.warn("X contains a zero vector, will not run cosine affinity.")
+
+        label_init = self.label_init
+        if label_init is not None:
+            if label_init.size != X.shape[0]:
+                raise ValueError(
+                    "n_samples must be the same as the length of label_init"
+                )
+
+        param_grid = dict(
+            affinity=affinity,
+            linkage=linkage,
+            covariance_type=covariance_type,
+            n_components=range(self.min_components, self.max_components + 1),
+        )
+
+        param_grid = list(ParameterGrid(param_grid))
+        param_grid_ag, param_grid = _process_paramgrid(
+            param_grid, self.n_init, self.label_init
+        )
+
+        seeds = random_state.randint(np.iinfo(np.int32).max, size=len(param_grid))
+
+        n_samples = X.shape[0]
+        if self.max_agglom_size is None or n_samples <= self.max_agglom_size:
+            X_subset = X
+        else:  # if dataset is huge, agglomerate a subset
+            subset_idxs = random_state.choice(
+                np.arange(0, n_samples), self.max_agglom_size
+            )
+            X_subset = X[subset_idxs, :]
+
+        ag_labels = []
+        if self.label_init is None:
+            for p_ag in param_grid_ag:
+                if p_ag["affinity"] != "none":
+                    agg = AgglomerativeClustering(
+                        n_clusters=self.min_components, **p_ag
+                    )
+                    agg.fit(X_subset)
+                    hierarchical_labels = _hierarchical_labels(
+                        agg.children_, self.min_components, self.max_components
+                    )
+                    ag_labels.append(hierarchical_labels)
+
+        def _fit_for_data(ag_params, gm_params, seed):
+            n_clusters = gm_params["n_components"]
+            if (ag_params["affinity"] != "none") and (self.label_init is None):
+                index = param_grid_ag.index(ag_params)
+                agg_clustering = ag_labels[index][:, n_clusters - self.min_components]
+            else:
+                agg_clustering = []
+            return self._fit_cluster(
+                X, X_subset, y, ag_params, gm_params, agg_clustering, seed
+            )
+
+        if parse_version(joblib.__version__) < parse_version("0.12"):
+            parallel_kwargs = {"backend": "threading"}
+        else:
+            parallel_kwargs = {"prefer": "threads"}
+
+        results = Parallel(n_jobs=self.n_jobs, verbose=self.verbose, **parallel_kwargs)(
+            delayed(_fit_for_data)(ag_params, gm_params, seed)
+            for (ag_params, gm_params), seed in zip(param_grid, seeds)
+        )
+        best_criter = [result.criterion for result in results]
+        # select the best model randomly in case there is a tie
+        best_idx = random_state.choice(np.where(best_criter == np.min(best_criter))[0])
```

Comment:
Naive question: wouldn't it be better to select the model with the least parameters if the BIC is the same?

Reply:
I think that's better. Just changed to it!